transaction_verifier.go

```diff
@@ -0,0 +1,148 @@
+		index >>= 1
+	}
+
+	return directions, nil
+}
+
+func (t *TransactionVerifier) computeVectorCommitmentMerkleRoot(leaf stateprooftypes.GenericDigest, leafIndex uint64, proof []byte, treeDepth uint64, hashFunc hash.Hash) (stateprooftypes.GenericDigest, error) {
+	if len(proof) == 0 && treeDepth == 0 {
+		return leaf, nil
+	}
+
+	nodeSize := uint64(hashFunc.Size())
+	if treeDepth*nodeSize != uint64(len(proof)) {
+		return stateprooftypes.GenericDigest{}, ErrProofLengthTreeDepthMismatch
+	}
+
+	positions, err := t.getVectorCommitmentPositions(leafIndex, treeDepth)
+	if err != nil {
+		return stateprooftypes.GenericDigest{}, err
+	}
+
+	currentNodeHash := leaf
+	for currentLevel := uint64(0); currentLevel < treeDepth; currentLevel++ {
+		siblingIndex := currentLevel * nodeSize
+		siblingHash := proof[siblingIndex : siblingIndex+nodeSize]
+
+		nextNode := []byte(stateprooftypes.MerkleArrayNode)
+		switch positions[currentLevel] {
+		case left:
+			nextNode = append(append(nextNode, currentNodeHash...), siblingHash...)
+		case right:
+			nextNode = append(append(nextNode, siblingHash...), currentNodeHash...)
+		default:
+			return []byte{}, ErrInvalidPosition
+		}
+
+		currentNodeHash = stateprooftypes.HashBytes(hashFunc, nextNode)
+	}
+
+	return currentNodeHash, nil
+}
+
+// VerifyTransaction receives a hashed transaction, a proof to compute the transaction's commitment, a proof
+// to compute the light block header's commitment, and an expected commitment to compare to.
+// Verification works as follows:
+//	1. Compute the transaction's vector commitment using the provided transaction proof.
+//	2. Build a candidate light block header using the computed vector commitment.
+// 	3. Compute the candidate light block header's vector commitment using the provided light block header proof.
+// 	4. Verify that the computed candidate vector commitment matches the expected vector commitment.
+func (t *TransactionVerifier) VerifyTransaction(transactionId types.Digest, transactionProofResponse models.ProofResponse, lightBlockHeaderProofResponse models.LightBlockHeaderProof, confirmedRound types.Round, seed stateprooftypes.Seed, blockIntervalCommitment types.Digest) error {
+	hashFunc, err := stateprooftypes.UnmarshalHashFunc(transactionProofResponse.Hashtype)
+	if err != nil {
+		return err
+	}
+
+	var stibHashDigest types.Digest
+	copy(stibHashDigest[:], transactionProofResponse.Stibhash[:])
+	transactionLeaf := t.getTransactionLeaf(transactionId, stibHashDigest, hashFunc)
+	transactionProofRoot, err := t.computeVectorCommitmentMerkleRoot(transactionLeaf, transactionProofResponse.Idx,
+		transactionProofResponse.Proof, transactionProofResponse.Treedepth, hashFunc)
+
+	if err != nil {
+		return err
+	}
+
+	// We build the candidate light block header using the computed transactionProofRoot, hash and verify it.
+	candidateLightBlockHeaderLeaf := t.getLightBlockHeaderLeaf(confirmedRound, transactionProofRoot, seed, hashFunc)
```

Comment:
Should directly use Sha256 here, and not use the same function from the transaction, I think